src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java

```diff
@@ -354,36 +354,39 @@ public ReadPileup getOverlappingFragmentFilteredPileup() {
      *
      * @return the newly filtered pileup
      */
-    public ReadPileup getOverlappingFragmentFilteredPileup(boolean discardDiscordant, Comparator<PileupElement> tieBreaker) {
-        Map<String, PileupElement> filteredPileup = new HashMap<String, PileupElement>();
-        Set<String> readNamesDeleted = new HashSet<>();
+    public ReadPileup getOverlappingFragmentFilteredPileup(boolean discardDiscordant, Comparator<PileupElement> tieBreaker, SAMFileHeader header) {
+        List<PileupElement> filteredPileupList = new ArrayList<PileupElement>();
 
-        for (PileupElement p : this) {
-            String readName = p.getRead().getName();
+        for( Map.Entry<String, ReadPileup> pileup : this.splitBySample(header, null).entrySet()) {
+            Map<String, PileupElement> filteredPileup = new HashMap<String, PileupElement>();
```

Comment:
This code looks right.  It's a little complicated though, so to help improve readability I might extract everything in the outer loop into a private function that returns the collection of PileupElements.  

So it could read like:
```
 for (ReadPileup pileup : this.splitBySample(header, null).values()) {
            Collection<PileupElement> elements = filterSingleSampleForOverlaps(pileup, tieBreaker, discardDiscordant);
            filteredPileupList.addAll(elements);
}
```

There's a bunch of extra work being done here to support the multisample case, if it is slowing down the tool too much we could definitely optimize it.  But I've already violated the prime directive of software engineering and told you to do a silly optimization that probably does nothing, so... lets wait and see what shows up on the profile...